ci/github-script/commits: allow reason for not cherry-picking

wolfgangwalther opened the pull request:
The first commit is from #433370 to avoid conflicts.

This came out of a discussion with @SigmaSquadron in #431344.

This change allows giving a reason via footer of the commit message for why this commit is not cherry-picked. This avoids having to "explain" the automated review comment afterwards - instead, this explanation can be given immediately when writing that commit.

For example, for an update of `xen` on the stable branch, this could be:

```
xen: 4.19.3-unstable-2025-07-09 -> 4.19.3

[... commit message ...]

No-Cherry-Pick: unstable is on a different minor version
```

This would then be shown as part of the automated review. The severity of this will be downgraded from "warning" to "important". We still treat the review as "changes requested", because it would be very complicated and noisy to handle two different categories of reviews, some with requested changes and some with comments only.

An alternative would be to not show this review at all. However, given that the reviewers expectation on backports should already be "if it's not a clean backport, the automated review will tell me what to look at", it seems better to show these and have the committer confirm by dismissing the review. Otherwise we risk merging actually unreviewed commits.

## Things done

Only tested for basic syntax errors, so far, I'd like to discuss the idea in general first.

- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
ci/github-script/commits: allow reason for not cherry-picking

This change allows giving a reason via footer of the commit message for
why this commit is not cherry-picked. This avoids having to "explain"
the automated review comment afterwards - instead, this explanation can
be given immediately when writing that commit.

For example, for an update of `xen` on the stable branch, this could be:

```
xen: 4.19.3-unstable-2025-07-09 -> 4.19.3

[... commit message ...]

Not-cherry-picked-because: unstable is on a different minor version
```

This would then be shown as part of the automated review. The severity
of this will be downgraded from "warning" to "important". We still treat
the review as "changes requested", because it would be very complicated
and noisy to handle two different categories of reviews, some with
requested changes and some with comments only.

An alternative would be to not show this review at all. However, given
that the reviewers expectation on backports should already be "if it's
not a clean backport, the automated review will tell me what to look
at", it seems better to show these and have the committer confirm by
dismissing the review. Otherwise we risk merging actually unreviewed
commits.

SigmaSquadron reviewed (approved):
> We still treat the review as "changes requested", because it would be very complicated and noisy to handle two different categories of reviews, some with requested changes and some with comments only.

Bit unfortunate, but seems good to me!

> An alternative would be to not show this review at all.

This seems like it could easily be abused. I'm okay with the bot posting a short message[^1] copy-pasting my reason why the commit isn't cherry-picked. We need to be notified if the commits aren't in `master`.

---

The code changes look okay at a glance to me, but JavaScript is beyond my realm of capabilities. I couldn't tell you if there are bugs hiding in the changes. The overall design looks excellent, though!

[^1]: I'm assuming the bot won't write the full message. Everything below should be omitted from the `No-Cherry-Picks` message: <br/><blockquote>Please follow the [backporting guidelines](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#how-to-backport-pull-requests) and cherry-pick with the -x flag. This requires changes to the unstable master and staging branches first, before backporting them. Occasionally, it is not possible to cherry-pick exactly the same patch. This most frequently happens when resolving merge conflicts or when updating minor versions of packages which have already advanced to the next major on unstable. If you need to merge this PR despite the warnings, please [dismiss](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/dismissing-a-pull-request-review) this review shortly before merging. Hint: The full diffs are also available in the [runner logs](https://github.com/NixOS/nixpkgs/actions/runs/16767118617/job/47474471092?pr=431344) with slightly better highlighting.<blockquote/>

SigmaSquadron commented:
I mean, ideally that would be removed, since if you're committing with No-Cherry-Picks, you don't want instructions on how to... cherry-pick.

wolfgangwalther committed:
ci/github-script/commits: clarify comments

Splits the "occasionally" case into two, depending on whether the commit
has a diff or was not cherry-picked at all. Prepares the next commit,
where these are conditionally shown only.

wolfgangwalther committed:
ci/github-script/commits: conditionally show comments

This only shows *some* of the additional hints, depending on what the
checks resulted in. Should hopefully reduce confusion a bit.

MattSturgeon reviewed (commented):
Don't see any major issues on my first scan over.

I believe the first commit is merged now, so can it be dropped in a rebase?

A couple minor questions, otherwise LGTM.

philiptaron reviewed (approved):
I'm fine with this. My intuition is that it'll be used on a small fraction of backport / PRs to the release branch and the status quo ("oh yeah, that's fine, ignore the rules for this one") will continue. As a result, the value delivered here is low, in my estimation.

I do appreciate the trailer `No-Cherry-Pick`. Bikeshedding, it could be named `Not-cherry-picked-because` or `Release-only-commit-reason` or `Merge-rule-exception`. Note that most trailers have an initial capital and then all lowercase.
